File: mf2/php/src/RelativeTimeCore.php
<?php

declare(strict_types=1);

namespace Mojito\MessageFormat2;

final class RelativeTimeCore
{
    private const DEFAULT_LOCALE = 'en';
    private const MAX_LOCALE_LENGTH = 256;
    private const MAX_OPTION_LENGTH = 256;
    private const MAX_OPERAND_LENGTH = 256;
    private const MAX_RELATIVE_TIME_QUANTITY = 1_000_000_000;
    private const DECIMAL_NUMBER_PATTERN = '/^-?(?:0|[1-9]\d*)(?:\.\d+)?(?:[eE][+-]?\d+)?$/';
    private const STYLE_LONG = 'long';
    private const STYLE_SHORT = 'short';
    private const STYLE_NARROW = 'narrow';
    private const NUMERIC_ALWAYS = 'always';
    private const NUMERIC_AUTO = 'auto';
    private const POLICY_PRECISE = 'precise';
    private const POLICY_COMPACT = 'compact';
    private const POLICY_CHAT = 'chat';
    private const UNIT_AUTO = 'auto';
    private const UNIT_SECONDS = [
        'second' => 1,
        'minute' => 60,
        'hour' => 3_600,
        'day' => 86_400,
        'week' => 604_800,
        'month' => 2_592_000,
        'quarter' => 7_776_000,
        'year' => 31_536_000,
    ];
    private const POLICIES = [
        self::POLICY_PRECISE => [
            [60, 'second'],
            [3_600, 'minute'],
            [86_400, 'hour'],
            [604_800, 'day'],
            [2_592_000, 'week'],
            [31_536_000, 'month'],
            [INF, 'year'],
        ],
        self::POLICY_COMPACT => [
            [60, 'second'],
            [3_600, 'minute'],
            [86_400, 'hour'],
            [INF, 'day'],
        ],
        self::POLICY_CHAT => [
            [45, 'second'],
            [2_700, 'minute'],
            [79_200, 'hour'],
            [604_800, 'day'],
            [INF, 'week'],
        ],
    ];

    private function __construct(
        private array $localeMap,
        private array $patternSets,
    ) {
    }

    public static function create(array $data): self
    {
        $localeMap = self::prepareLocaleMap($data['localeMap'] ?? null);
        $rawPatternSets = $data['patternSets'] ?? null;
        if ($localeMap === null || !is_array($rawPatternSets) || $rawPatternSets === []) {
            throw new MF2Error('missing-locale-data', 'Relative-time core data has an unsupported shape.');
        }

        $patternSets = [];
        foreach ($rawPatternSets as $item) {
            if (is_array($item) && is_string($item['id'] ?? null) && $item['id'] !== '' && is_array($item['data'] ?? null) && $item['data'] !== []) {
                $patternSets[$item['id']] = $item['data'];
            }
        }
        if ($patternSets === []) {
            throw new MF2Error('missing-locale-data', 'Relative-time core data has an unsupported shape.');
        }
        return new self($localeMap, $patternSets);
    }

    private static function prepareLocaleMap(mixed $localeMap): ?array
    {
        if (!is_array($localeMap) || $localeMap === []) {
            return null;
        }
        foreach ($localeMap as $locale => $setId) {
            if (!is_string($locale) || !is_string($setId)) {
                return null;
            }
        }
        return $localeMap;
    }

    public static function registry(array $data): FunctionRegistry
    {
        return self::create($data)->functionRegistry();
    }

    public static function format(mixed $value, array $data, array $options = []): string
    {
        return self::create($data)->formatValue($value, $options);
    }

    public static function formatToParts(mixed $value, array $data, array $options = []): array
    {
        return self::create($data)->formatValueToParts($value, $options);
    }

    public function functionRegistry(): FunctionRegistry
    {
        return FunctionRegistry::portable()
            ->withFunction('relativeTime', fn(array $call): string => $this->formatCall($call));
    }

    public function formatValue(mixed $value, array $options = []): string
    {
        $style = self::optionOneOf(self::stringOption($options['style'] ?? self::STYLE_SHORT, 'style'), [self::STYLE_LONG, self::STYLE_SHORT, self::STYLE_NARROW], 'style');
        $numeric = self::optionOneOf(self::stringOption($options['numeric'] ?? self::NUMERIC_ALWAYS, 'numeric'), [self::NUMERIC_ALWAYS, self::NUMERIC_AUTO], 'numeric');
        $policy = self::optionOneOf(self::stringOption($options['policy'] ?? self::POLICY_PRECISE, 'policy'), [self::POLICY_PRECISE, self::POLICY_COMPACT, self::POLICY_CHAT], 'policy');
        $unit = self::optionOneOf(self::stringOption($options['unit'] ?? self::UNIT_AUTO, 'unit'), array_merge([self::UNIT_AUTO], array_keys(self::UNIT_SECONDS)), 'unit');
        $seconds = self::parseFiniteNumber($value);
        $selectedUnit = $unit === self::UNIT_AUTO ? self::selectUnit($seconds, $policy) : $unit;
        $quantity = self::quantity($seconds, $selectedUnit);
        $locale = self::localeOption($options['locale'] ?? self::DEFAULT_LOCALE);
        if ($locale === '') {
            $locale = self::DEFAULT_LOCALE;
        }

        if (self::useRelativeZero($policy, $numeric, $seconds)) {
            $relative = $this->relativeTerm($locale, $style, $selectedUnit, '0');
            if ($relative !== null) {
                return $relative;
            }
        }
        if ($numeric === self::NUMERIC_AUTO) {
            $offset = self::relativeOffset($seconds, $selectedUnit, $quantity);
            if ($offset !== null) {
                $relative = $this->relativeTerm($locale, $style, $selectedUnit, $offset);
                if ($relative !== null) {
                    return $relative;
                }
            }
        }

        $direction = self::isNegativeRelativeTime($seconds) ? 'past' : 'future';
        $category = Internal\select_cardinal($locale, $quantity);
        $pattern = $this->relativeTimePattern($locale, $style, $selectedUnit, $direction, $category);
        return str_replace('{0}', (string) $quantity, $pattern);
    }

    public function formatValueToParts(mixed $value, array $options = []): array
    {
        return [['type' => 'text', 'value' => $this->formatValue($value, $options)]];
    }

    private function formatCall(array $call): string
    {
        return $this->formatValue($call['rawValue'] ?? $call['value'], [
            'locale' => $call['locale'] ?? self::DEFAULT_LOCALE,
            'style' => self::callOption($call, 'style', self::STYLE_SHORT),
            'numeric' => self::callOption($call, 'numeric', self::NUMERIC_ALWAYS),
            'policy' => self::callOption($call, 'policy', self::POLICY_PRECISE),
            'unit' => self::callOption($call, 'unit', self::UNIT_AUTO),
        ]);
    }

    private static function parseFiniteNumber(mixed $value): float
    {
        if (is_bool($value) || $value === null || is_array($value)) {
            throw MF2Error::badOperand('Relative-time core requires a finite numeric value.');
        }
        if (is_object($value) && !method_exists($value, '__toString')) {
            throw MF2Error::badOperand('Relative-time core requires a finite numeric value.');
        }
        $text = trim(self::stringOperand($value));
        if ($text === '') {
            throw MF2Error::badOperand('Relative-time core requires a finite numeric value.');
        }
        if (strlen($text) > self::MAX_OPERAND_LENGTH) {
            throw MF2Error::badOperand('Relative-time core requires a finite numeric value.');
        }
        if (preg_match(self::DECIMAL_NUMBER_PATTERN, $text) !== 1) {
            throw MF2Error::badOperand('Relative-time core requires a finite numeric value.');
        }
        $parsed = (float) $text;
        if (!is_numeric($text) || !is_finite($parsed)) {
            throw MF2Error::badOperand('Relative-time core requires a finite numeric value.');
        }
        return $parsed;
    }

    private static function optionOneOf(string $value, array $allowed, string $name): string
    {
        if (!in_array($value, $allowed, true)) {
            throw MF2Error::badOption("{$name} must be one of " . implode(', ', $allowed) . '.');
        }
        return $value;
    }

    private static function stringOperand(mixed $value): string
    {
        try {
            return (string) $value;
        } catch (\Throwable) {
            throw MF2Error::badOperand('Relative-time core requires a finite numeric value.');
        }
    }

    private static function stringOption(mixed $value, string $name): string
    {
        if (is_array($value)) {
            throw MF2Error::badOption("{$name} must be coercible to a string.");
        }
        try {
            $text = (string) $value;
        } catch (\Throwable) {
            throw MF2Error::badOption("{$name} must be coercible to a string.");
        }
        if (strlen($text) > self::MAX_OPTION_LENGTH) {
            throw MF2Error::badOption("{$name} must not exceed 256 characters.");
        }
        return $text;
    }

    private static function localeOption(mixed $value): string
    {
        $locale = self::stringOption($value, 'locale');
        if (strlen($locale) > self::MAX_LOCALE_LENGTH) {
            throw MF2Error::badOption('locale must not exceed 256 characters.');
        }
        return $locale;
    }

    private static function selectUnit(float $seconds, string $policy): string
    {
        $absolute = abs($seconds);
        foreach (self::POLICIES[$policy] as [$upper, $unit]) {
            if ($absolute < $upper) {
                return $unit;
            }
        }
        return 'year';
    }

    private static function quantity(float $seconds, string $unit): int
    {
        $absolute = abs($seconds);
        if ($absolute === 0.0) {
            return 0;
        }
        $quantity = max(1.0, floor($absolute / self::UNIT_SECONDS[$unit] + 0.5));
        if ($quantity > self::MAX_RELATIVE_TIME_QUANTITY) {
            throw MF2Error::badOperand('Relative-time core quantity is outside the supported range.');
        }
        return (int) $quantity;
    }

    private static function useRelativeZero(string $policy, string $numeric, float $seconds): bool
    {
        return $policy === self::POLICY_CHAT && $numeric === self::NUMERIC_AUTO && abs($seconds) < 45;
    }

    private static function relativeOffset(float $seconds, string $unit, int $quantity): ?string
    {
        if ($quantity === 0) {
            return '0';
        }
        if (abs($seconds) !== ((float) $quantity) * self::UNIT_SECONDS[$unit]) {
            return null;
        }
        return self::isNegativeRelativeTime($seconds) ? '-' . $quantity : (string) $quantity;
    }

    private static function isNegativeRelativeTime(float $seconds): bool
    {
        return $seconds < 0 || ($seconds === 0.0 && json_encode($seconds) === '-0');
    }

    private function relativeTerm(string $locale, string $style, string $unit, string $offset): ?string
    {
        $relative = $this->unitData($locale, $style, $unit)['relative'] ?? [];
        return is_array($relative) && is_string($relative[$offset] ?? null) ? $relative[$offset] : null;
    }

    private function relativeTimePattern(string $locale, string $style, string $unit, string $direction, string $category): string
    {
        $patterns = $this->unitData($locale, $style, $unit)[$direction] ?? null;
        if (is_array($patterns)) {
            $pattern = $patterns[$category] ?? $patterns['other'] ?? null;
            if (is_string($pattern)) {
                return $pattern;
            }
        }
        throw new MF2Error('missing-locale-data', "Missing relative-time pattern for {$locale}/{$style}/{$unit}/{$direction}.");
    }

    private function unitData(string $locale, string $style, string $unit): array
    {
        $patternSet = $this->patternSetFor($locale);
        $unitData = $patternSet[$style][$unit] ?? null;
        if (is_array($unitData)) {
            return $unitData;
        }
        throw new MF2Error('missing-locale-data', "Missing relative-time unit data for {$locale}/{$style}/{$unit}.");
    }

    private function patternSetFor(string $locale): array
    {
        foreach (Internal\locale_lookup_chain($locale) as $candidate) {
            $setId = $this->localeMap[$candidate] ?? null;
            if (!is_string($setId)) {
                continue;
            }
            $patternSet = $this->patternSets[$setId] ?? null;
            if (!is_array($patternSet)) {
                throw new MF2Error('missing-locale-data', "Missing relative-time pattern set {$setId}.");
            }
            return $patternSet;
        }
        throw new MF2Error('missing-locale-data', "Missing relative-time locale data for {$locale}.");
    }

    private static function callOption(array $call, string $name, mixed $fallback = null): mixed
    {
        $optionValue = $call['optionValue'] ?? static fn(string $name, mixed $fallback): mixed => $fallback;
        return $optionValue($name, $fallback);
    }
}

File: mf2/python/src/mojito_mf2/relative_time_core.py
from __future__ import annotations

from dataclasses import dataclass
import math
import re
from typing import Any, Mapping

from ._cldr_plural_rules import select_cardinal
from ._locale_key import locale_lookup_chain
from .errors import MF2Error
from .functions import FunctionCall, FunctionRegistry

__all__ = [
    "format_relative_time_core",
    "format_relative_time_core_to_parts",
    "relative_time_core_function_registry",
]

_DEFAULT_LOCALE = "en"
_MAX_LOCALE_LENGTH = 256
_MAX_OPTION_LENGTH = 256
_MAX_OPERAND_LENGTH = 256
_MAX_RELATIVE_TIME_QUANTITY = 1_000_000_000
_DECIMAL_NUMBER_RE = re.compile(r"^-?(?:0|[1-9][0-9]*)(?:\.[0-9]+)?(?:[eE][+-]?[0-9]+)?$")
_STYLE_VALUES = {"long", "short", "narrow"}
_NUMERIC_VALUES = {"always", "auto"}
_POLICY_VALUES = {"precise", "compact", "chat"}
_UNIT_SECONDS = {
    "second": 1,
    "minute": 60,
    "hour": 3_600,
    "day": 86_400,
    "week": 604_800,
    "month": 2_592_000,
    "quarter": 7_776_000,
    "year": 31_536_000,
}
_UNIT_VALUES = {"auto", *_UNIT_SECONDS}
_POLICIES = {
    "precise": (
        (60, "second"),
        (3_600, "minute"),
        (86_400, "hour"),
        (604_800, "day"),
        (2_592_000, "week"),
        (31_536_000, "month"),
        (math.inf, "year"),
    ),
    "compact": (
        (60, "second"),
        (3_600, "minute"),
        (86_400, "hour"),
        (math.inf, "day"),
    ),
    "chat": (
        (45, "second"),
        (2_700, "minute"),
        (79_200, "hour"),
        (604_800, "day"),
        (math.inf, "week"),
    ),
}


@dataclass(frozen=True)
class _RelativeTimeData:
    locale_map: Mapping[str, str]
    pattern_sets: Mapping[str, Mapping[str, Any]]


def format_relative_time_core(
    value: Any,
    *,
    data: Mapping[str, Any],
    locale: str = _DEFAULT_LOCALE,
    style: str = "short",
    numeric: str = "always",
    policy: str = "precise",
    unit: str = "auto",
) -> str:
    prepared = _prepare_data(data)
    return _format_relative_time_prepared(
        value,
        data=prepared,
        locale=locale,
        style=style,
        numeric=numeric,
        policy=policy,
        unit=unit,
    )


def format_relative_time_core_to_parts(value: Any, **options: Any) -> list[dict[str, str]]:
    return [{"type": "text", "value": format_relative_time_core(value, **options)}]


def relative_time_core_function_registry(data: Mapping[str, Any]) -> FunctionRegistry:
    prepared = _prepare_data(data)
    return FunctionRegistry.portable().with_function(
        "relativeTime",
        lambda call: _format_call_relative_time(call, prepared),
    )


def _format_call_relative_time(call: FunctionCall, data: _RelativeTimeData) -> str:
    value = call.raw_value if call.raw_value is not None else call.value
    options = {
        "data": data,
        "locale": call.locale,
        "style": call.option_value("style", "short") or "short",
        "numeric": call.option_value("numeric", "always") or "always",
        "policy": call.option_value("policy", "precise") or "precise",
        "unit": call.option_value("unit", "auto") or "auto",
    }
    try:
        return _format_relative_time_prepared(value, **options)
    except MF2Error as error:
        source_value = _relative_time_source_value(call.inherited_source)
        if error.code != "bad-operand" or source_value is None:
            raise
        return _format_relative_time_prepared(source_value, **options)


def _relative_time_source_value(source: object | None) -> str | None:
    while source is not None:
        function = getattr(source, "function")
        if function.get("name") in {
            "number",
            "integer",
            "percent",
            "offset",
            "currency",
            "relativeTime",
        }:
            return getattr(source, "value")
        source = getattr(source, "inherited_source")
    return None


def _format_relative_time_prepared(
    value: Any,
    *,
    data: _RelativeTimeData,
    locale: str,
    style: str,
    numeric: str,
    policy: str,
    unit: str,
) -> str:
    locale = _coerce_string_option(locale, "locale")
    _validate_locale_length(locale)
    style = _option_one_of(style, _STYLE_VALUES, "style")
    numeric = _option_one_of(numeric, _NUMERIC_VALUES, "numeric")
    policy = _option_one_of(policy, _POLICY_VALUES, "policy")
    unit = _option_one_of(unit, _UNIT_VALUES, "unit")
    seconds = _parse_finite_number(value)
    selected_unit = _select_unit(seconds, policy) if unit == "auto" else unit
    quantity = _relative_time_quantity(seconds, selected_unit)

    if _use_relative_zero(policy, numeric, seconds):
        relative = _relative_term(data, locale, style, selected_unit, "0")
        if relative is not None:
            return relative

    if numeric == "auto":
        offset = _relative_offset(seconds, selected_unit, quantity)
        if offset is not None:
            relative = _relative_term(data, locale, style, selected_unit, offset)
            if relative is not None:
                return relative

    direction = "past" if _is_negative_relative_time(seconds) else "future"
    category = select_cardinal(quantity, locale)
    pattern = _relative_time_pattern(data, locale, style, selected_unit, direction, category)
    return pattern.replace("{0}", str(quantity))


def _prepare_data(data: Mapping[str, Any]) -> _RelativeTimeData:
    if not isinstance(data, Mapping):
        raise MF2Error("missing-locale-data", "Relative-time core data has an unsupported shape.")
    locale_map = data.get("localeMap")
    pattern_sets = data.get("patternSets")
    if (
        not isinstance(locale_map, Mapping)
        or not locale_map
        or not isinstance(pattern_sets, list)
        or not pattern_sets
    ):
        raise MF2Error("missing-locale-data", "Relative-time core data has an unsupported shape.")

    decoded_locale_map: dict[str, str] = {}
    for locale, set_id in locale_map.items():
        if not isinstance(locale, str) or not isinstance(set_id, str):
            raise MF2Error("missing-locale-data", "Relative-time core data has an unsupported shape.")
        decoded_locale_map[locale] = set_id

    decoded_pattern_sets: dict[str, Mapping[str, Any]] = {}
    for item in pattern_sets:
        if (
            isinstance(item, Mapping)
            and isinstance(item.get("id"), str)
            and item["id"] != ""
            and isinstance(item.get("data"), Mapping)
            and bool(item["data"])
        ):
            decoded_pattern_sets[item["id"]] = item["data"]
    if not decoded_pattern_sets:
        raise MF2Error("missing-locale-data", "Relative-time core data has an unsupported shape.")
    return _RelativeTimeData(locale_map=decoded_locale_map, pattern_sets=decoded_pattern_sets)


def _parse_finite_number(value: Any) -> float:
    if isinstance(value, bool) or value is None:
        raise MF2Error("bad-operand", "Relative-time core requires a finite numeric value.")
    text = _coerce_string_operand(value).strip()
    if not text:
        raise MF2Error("bad-operand", "Relative-time core requires a finite numeric value.")
    if len(text) > _MAX_OPERAND_LENGTH:
        raise MF2Error("bad-operand", "Relative-time core requires a finite numeric value.")
    if not _DECIMAL_NUMBER_RE.fullmatch(text):
        raise MF2Error("bad-operand", "Relative-time core requires a finite numeric value.")
    try:
        parsed = float(text)
    except ValueError as error:
        raise MF2Error("bad-operand", "Relative-time core requires a finite numeric value.") from error
    if not math.isfinite(parsed):
        raise MF2Error("bad-operand", "Relative-time core requires a finite numeric value.")
    return parsed


def _coerce_string_operand(value: Any) -> str:
    try:
        return str(value)
    except Exception as error:
        raise MF2Error("bad-operand", "Relative-time core requires a finite numeric value.") from error


def _coerce_string_option(value: Any, name: str) -> str:
    try:
        text = str(value)
    except Exception as error:
        raise MF2Error("bad-option", f"{name} must be coercible to a string.") from error
    if len(text) > _MAX_OPTION_LENGTH:
        raise MF2Error("bad-option", f"{name} must not exceed 256 characters.")
    return text


def _validate_locale_length(locale: str) -> None:
    if len(locale) > _MAX_LOCALE_LENGTH:
        raise MF2Error("bad-option", "locale must not exceed 256 characters.")


def _option_one_of(value: Any, allowed: set[str], name: str) -> str:
    text = _coerce_string_option(value, name)
    if text not in allowed:
        raise MF2Error("bad-option", f"{name} must be one of {', '.join(sorted(allowed))}.")
    return text


def _select_unit(seconds: float, policy: str) -> str:
    absolute = abs(seconds)
    for upper, unit in _POLICIES[policy]:
        if absolute < upper:
            return unit
    return "year"


def _relative_time_quantity(seconds: float, unit: str) -> int:
    absolute = abs(seconds)
    if absolute == 0:
        return 0
    quantity = max(1, math.floor(absolute / _UNIT_SECONDS[unit] + 0.5))
    if quantity > _MAX_RELATIVE_TIME_QUANTITY:
        raise MF2Error("bad-operand", "Relative-time core quantity is outside the supported range.")
    return quantity


def _use_relative_zero(policy: str, numeric: str, seconds: float) -> bool:
    return policy == "chat" and numeric == "auto" and abs(seconds) < 45


def _relative_offset(seconds: float, unit: str, quantity: int) -> str | None:
    if quantity == 0:
        return "0"
    if abs(seconds) != quantity * _UNIT_SECONDS[unit]:
        return None
    return f"-{quantity}" if _is_negative_relative_time(seconds) else str(quantity)


def _is_negative_relative_time(seconds: float) -> bool:
    return seconds < 0 or math.copysign(1.0, seconds) < 0


def _relative_term(
    data: _RelativeTimeData,
    locale: str,
    style: str,
    unit: str,
    offset: str,
) -> str | None:
    relative = _relative_unit_data(data, locale, style, unit).get("relative", {})
    if isinstance(relative, Mapping):
        value = relative.get(offset)
        return value if isinstance(value, str) else None
    return None


def _relative_time_pattern(
    data: _RelativeTimeData,
    locale: str,
    style: str,
    unit: str,
    direction: str,
    category: str,
) -> str:
    patterns = _relative_unit_data(data, locale, style, unit).get(direction)
    if isinstance(patterns, Mapping):
        pattern = patterns.get(category, patterns.get("other"))
        if isinstance(pattern, str):
            return pattern
    raise MF2Error(
        "missing-locale-data",
        f"Missing relative-time pattern for {locale}/{style}/{unit}/{direction}.",
    )


def _relative_unit_data(
    data: _RelativeTimeData,
    locale: str,
    style: str,
    unit: str,
) -> Mapping[str, Any]:
    pattern_set = _pattern_set_for(data, locale)
    style_data = pattern_set.get(style)
    unit_data = style_data.get(unit) if isinstance(style_data, Mapping) else None
    if isinstance(unit_data, Mapping):
        return unit_data
    raise MF2Error(
        "missing-locale-data",
        f"Missing relative-time unit data for {locale}/{style}/{unit}.",
    )


def _pattern_set_for(data: _RelativeTimeData, locale: str) -> Mapping[str, Any]:
    for candidate in locale_lookup_chain(locale or _DEFAULT_LOCALE):
        set_id = data.locale_map.get(candidate)
        if set_id is None:
            continue
        pattern_set = data.pattern_sets.get(set_id)
        if pattern_set is not None:
            return pattern_set
        raise MF2Error("missing-locale-data", f"Missing relative-time pattern set {set_id}.")
    raise MF2Error("missing-locale-data", f"Missing relative-time locale data for {locale}.")
